File: python/2176-Count Equal and Divisible Pairs in an Array.py
from collections import defaultdict

class Solution(object):
  def countPairs(self, nums, k):
    valToIdxs = defaultdict(list)
    for i, e in enumerate(nums):
      valToIdxs[e].append(i)

    if max(len(item) for item in valToIdxs.values()) < 2:
      return 0
    
    ret = 0
    for value, idxs in valToIdxs.items():
      for i, firstIdx in enumerate(idxs[:-1]):
        for j, secondIdx in enumerate(idxs):
          if firstIdx != secondIdx and firstIdx < secondIdx:
            if 0 == firstIdx * secondIdx % k:
              ret += 1

    return ret

File: cpp/2176-Count Equal and Divisible PAirs in an Array.cpp
#include <map>
#include <vector>
#include <iostream>

class Solution
{
public:
  int countPairs(vector<int>& nums, int k)
  {
    std::map<int, std::vector<int>> valsToIdxs;
    for(size_t i = 0; i < nums.size(); i++)
    {
      valsToIdxs[nums[i]].push_back(i);
    }

    int ret = 0;
    for(auto it = valsToIdxs.begin(); it != valsToIdxs.end(); it++)
    {
      for(size_t i = 0; i < it->second.size() - 1; i++)
      {
        for(size_t j = i + 1; j < it->second.size(); j++)
        {
          if(0 == it->second[i] * it->second[j] % k)
          {
            ret++;
          }
        }
      }
    }
    return ret;
  }
};